[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144794

Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)

[ghstack-poisoned]

kwen2501 commented:
`ncclCommInitRankScalable` accepts `m` unique IDs, where `m` is between 1 and `nRanks`. 
Those`m` unique IDs are generated by `m` root ranks -- one easy way to elect those roots is one root per `nRanks / m` ranks.

We'd need to perform `m` broadcasts to help those `m` roots broadcast their unique IDs to the world.

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
Thanks for the PR. Quick comment: shall we add a test case?
Will have a closer look tmr.

wconstab reviewed (commented):
what about tests?

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@wconstab lol, I was working on the test today, and now I updated it in the PR.

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2155,6 +2156,22 @@ void ProcessGroupNCCL::checkAndSetRemoteError() {
+  const int rmr = nRanks % nIds;
+  const int rpr = nRanks / nIds;
+  const int rlim = rmr * (rpr + 1);
+  if (rank < rlim) {
+    return 0;
+  } else {
+    return ((rank - rlim) / nIds) * nIds;
+  }
```
This is wrong, I need to think more on this one.

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
unit test passed when patching https://github.com/pytorch/pytorch/pull/142179.

fduwjj commented:
waiting for issue: https://github.com/pytorch/pytorch/issues/144768#issue-2787280299 and https://github.com/pytorch/pytorch/pull/142179.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2537,6 +2567,57 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
   }
 }
 
+void ProcessGroupNCCL::allgatherUniqueNCCLID(
+    int rootRank,
+    ncclUniqueId* ncclID,
+    std::vector<ncclUniqueId>& ncclIDs) {
```
nit: `allgatherUniqueNCCLIDs` (s at the end).

nit: document the arg fields.

fduwjj replied:
Looks like we don't usually do that in cpp for the arg doc..

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2537,6 +2567,57 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
+        reinterpret_cast<uint8_t*>(ncclID),
+        reinterpret_cast<uint8_t*>(ncclID) + NCCL_UNIQUE_ID_BYTES);
+    store_->set(storeKeys[rootRank], vec);
+  }
+  try {
+    int indx = 0;
+    for (auto& value : store_->multiGet(storeKeys)) {
+      TORCH_CHECK_WITH(
+          DistBackendError,
+          value.size() == NCCL_UNIQUE_ID_BYTES,
+          "Invalid size for ncclUniqueId");
+      std::memcpy(&ncclIDs[indx++], value.data(), value.size());
+    }
+  } catch (const std::exception& e) {
+    nlohmann::json json_vec = storeKeys;
+    std::string exceptionMsg = c10::str(
+        "[",
+        rank_,
+        "] is setting up NCCL communicators and "
+        "retrieving ncclUniqueId from roots via TCPStore by key '",
+        json_vec.dump(),
+        "', but got error: ");
+    C10_THROW_ERROR(
+        DistBackendError,
+        exceptionMsg + e.what() +
+            ". This may indicate a possible application crash on rank 0 or a network set up issue.");
+  } catch (...) {
+    nlohmann::json json_vec = storeKeys;
+    C10_THROW_ERROR(
+        DistBackendError,
+        c10::str(
+            "Unknown exception while [",
+            rank_,
+            "] is setting up NCCL communicators and "
+            "retrieving ncclUniqueIds from roots via TCPStore by key '",
+            json_vec.dump(),
+            "'",
+            ". This may indicate a possible application crash on rank 0 or a network set up issue."));
+  }
+}
```
nit: comments would be appreciated in this function.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2537,6 +2567,57 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
+        reinterpret_cast<uint8_t*>(ncclID) + NCCL_UNIQUE_ID_BYTES);
+    store_->set(storeKeys[rootRank], vec);
+  }
+  try {
+    int indx = 0;
+    for (auto& value : store_->multiGet(storeKeys)) {
+      TORCH_CHECK_WITH(
+          DistBackendError,
+          value.size() == NCCL_UNIQUE_ID_BYTES,
+          "Invalid size for ncclUniqueId");
+      std::memcpy(&ncclIDs[indx++], value.data(), value.size());
+    }
+  } catch (const std::exception& e) {
```
nit: shall we only wrap `store_->multiGet` with try-catch? 
The other part of the code does not seem relevant to the handling in catch.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2537,6 +2567,57 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
+      std::memcpy(&ncclIDs[indx++], value.data(), value.size());
+    }
+  } catch (const std::exception& e) {
+    nlohmann::json json_vec = storeKeys;
+    std::string exceptionMsg = c10::str(
+        "[",
+        rank_,
+        "] is setting up NCCL communicators and "
+        "retrieving ncclUniqueId from roots via TCPStore by key '",
+        json_vec.dump(),
+        "', but got error: ");
+    C10_THROW_ERROR(
+        DistBackendError,
+        exceptionMsg + e.what() +
+            ". This may indicate a possible application crash on rank 0 or a network set up issue.");
+  } catch (...) {
+    nlohmann::json json_vec = storeKeys;
+    C10_THROW_ERROR(
+        DistBackendError,
+        c10::str(
+            "Unknown exception while [",
+            rank_,
+            "] is setting up NCCL communicators and "
+            "retrieving ncclUniqueIds from roots via TCPStore by key '",
+            json_vec.dump(),
+            "'",
+            ". This may indicate a possible application crash on rank 0 or a network set up issue."));
+  }
```
nit: I wonder if there is a way to avoid duplicated code. (It may be hard, and I haven't done any research either)

fduwjj replied:
I thought about it. Unfortunately, since the logging message and API usage is different. I decide to not to reuse the existing code.

kwen2501 replied:
Sorry about the confusion. I meant dedup the two catch blocks here.

fduwjj replied:
I don't think so, this is literally to generate logs and throw exceptions..

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2678,6 +2759,32 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::initNCCLComm(
+  // accoring to
+  // https://github.com/pytorch/pytorch/pull/136789#discussion_r1779171615.
+  auto ranksPerRoot = getCvarInt(TORCH_NCCL_SCALABLE_INIT_RANKS_PER_ROOT, 128);
+  auto numRoots = (getSize() / ranksPerRoot) + (getSize() % ranksPerRoot > 0);
```
nit: `(getSize() + ranksPerRoot - 1) / ranksPerRoot`

fduwjj replied:
Oh I forgot this way...

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2678,6 +2759,32 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::initNCCLComm(
+  std::vector<ncclUniqueId> ncclIDs(numRoots);
+
+  if (!ncclComm) {
+    C10D_NCCL_CHECK(ncclGetUniqueId(&ncclID), std::nullopt);
```
Hmm, is this called on every rank? (should not be)

fduwjj replied:
I think I need some help on how to deal with P2P case.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2678,6 +2759,32 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::initNCCLComm(
+            .count() *
+        1000;
+    LOG(INFO) << logPrefix()
+              << "ProcessGroupNCCL broadcast unique ID through store took "
```
nit: "broadcast" -> "all-gather"

fduwjj replied:
Opps

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2537,6 +2567,57 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
+  for (size_t r = 0; r < ncclIDs.size(); r++) {
+    storeKeys.emplace_back("UniqueNCCLID:" + std::to_string(r));
+  }
+  if (rank_ == rootRank) {
+    auto vec = std::vector<uint8_t>(
+        reinterpret_cast<uint8_t*>(ncclID),
+        reinterpret_cast<uint8_t*>(ncclID) + NCCL_UNIQUE_ID_BYTES);
+    store_->set(storeKeys[rootRank], vec);
+  }
```
I wonder if `storeKeys[rootRank]` would overflow, because `storeKeys` has length of `numRoots`, but `if (rank_ == rootRank)` seems to suggest that `rootRank` can range from 0 to `world_size - 1`.

fduwjj replied:
technically yes, but in reality, we use 128 root per rank, do you still think it is possible? Or I can add a warning there?

kwen2501 replied:
I don't know what `rootRank` stands for, is my point.

fduwjj replied:
it actually should be rootIdx..

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2155,6 +2156,35 @@ void ProcessGroupNCCL::checkAndSetRemoteError() {
+// rmr = 10 % 3 = 1 (1 larger group),
+// rpr = 10 / 3 = 3 (base number of ranks per group).
+// rlim = 4
+// Output root:
+// For ranks [0, 1, 2, 3] is 0.
+// For ranks [4, 5, 6] is 1.
+// For ranks [7, 8, 9] is 2.
```
Are the output roots 0, 1, 2 absolute rank IDs?
This choice seems different from recommendation:
"For example, if 3 ncclUniqueIds are to be distributed accross 7 NCCL ranks, the first ncclUniqueId will be associated to ranks 0-2, while the others will be associated to ranks 3-4, and 5-6. This function will therefore return true on **rank 0, 3, and 5**, and false otherwise."

fduwjj replied:
oh, I think it gives out 0, 3 and 5 in this case..

kwen2501 reviewed (commented):
Overall LGTM. Very close to land, just need to verify / address some minor details.

fduwjj commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -253,6 +253,14 @@ def test_init_wo_backend_str(self):
+    @requires_nccl()
+    @skip_if_lt_x_gpu(1)
+    def test_scalable_init(self):
+        os.environ["TORCH_NCCL_SCALABLE_INIT_RANKS_PER_ROOT"] = "1"
```
Then how about `TORCH_NCCL_RANKS_PER_ROOT`?

kwen2501 replied:
sgtm

fduwjj commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -87,7 +87,33 @@ std::shared_ptr<NCCLComm> NCCLComm::create(
+          commIds.data(),
+          &config),
+      std::nullopt);
+  comm->ncclId_ = commIds[rank];
```
oh this should be the root rank not the actual rank... let me change that.

kwen2501 replied:
> this should be the root rank

hmm, actually, the first ID is what matters.

fduwjj replied:
yeah corrected it.

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -181,6 +182,12 @@ static std::vector<std::string> TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK =
     {"TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK",
      "NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK"};
 
+// Return the rank of root and index during NCCL scalable comm init.
+static std::pair<int, int> getRootRankAndIndex(
+    const int rank,
+    const int nRanks,
+    const int nIds);
+
```
No need to put this function in header file.

fduwjj replied:
oh you mean directly define in cpp? OK.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -18,6 +18,7 @@
 #include <mutex>
 #include <thread>
 #include <unordered_map>
+#include <utility>
```
what is this include for?

fduwjj replied:
for this one std::pair<int, int>.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2568,29 +2572,27 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
   for (size_t r = 0; r < ncclIDs.size(); r++) {
     storeKeys.emplace_back("UniqueNCCLID:" + std::to_string(r));
   }
   if (rank_ == rootRank) {
```
I was wondering, which `rootRank` do you pass to each rank, for this if testing.

fduwjj replied:
the rootRank belongs to that specific rank.

fduwjj replied:
For the example 0 - 10 and 3 nodes.

root rank for [0, 3] is 0
root rank for [4, 7] is 4
root rank for [8, 10] is 8.

And rank_ here is just the global rank.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2760,64 +2770,73 @@ std::shared_ptr<NCCLComm> ProcessGroupNCCL::initNCCLComm(
-                << "ProcessGroupNCCL broadcast unique ID through store took "
+                << "ProcessGroupNCCL all-gather unique IDs through store took "
                 << timerDeltaMs << " ms";
+      ncclComm = NCCLComm::create_scalable(
+          numRanks, rootRankAndIndex.second, rank, ncclIDs, options_->config);
```
nit: you may have to either:
- add `#ifdef` for `create_scalable`, with a `#else` throwing an internal error. 
OR
- declare `create_scalable` unconditionally in NCCLUtils.hpp.

fduwjj replied:
oh... no wonder It has errors.

kwen2501 commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -107,7 +108,7 @@ std::shared_ptr<NCCLComm> NCCLComm::create_scalable(
           &config),
       std::nullopt);
-  comm->ncclId_ = commIds[rank];
+  comm->ncclId_ = commIds[rootIdx];
```
As previously mentioned, the *first* NCCL ID matters more.
Also so that all ranks prints the same NCCL ID in their log when representing this comm.

fduwjj replied:
Are you saying that we should do `comm->ncclId_ = commIds[0];` instead?

kwen2501 replied:
Yes, probably.

kwen2501 reviewed (approved):
Very close to land. Some minor comments for the changes.

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

eqy commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -253,6 +253,14 @@ def test_init_wo_backend_str(self):
+    @requires_nccl()
+    @skip_if_lt_x_gpu(1)
+    def test_scalable_init(self):
+        os.environ["TORCH_NCCL_SCALABLE_INIT_RANKS_PER_ROOT"] = "1"
```
Should this be cleaned up after the test in case of interference with other tests?

eqy replied:
also note that this has been changed in the `.cpp` source but it has not been updated in the test

fduwjj replied:
Opps sorry.

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][NCCL] Implement ncclCommInitRankScalable (merging #136789)"


Try to land https://github.com/pytorch/pytorch/pull/136789/files on our end and fix any remaining issues.


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

atalman reviewed (approved):
lgtm

fduwjj commented:
@pytorchbot merge -i
